extensions/amp-subscriptions/0.1/platform-store.js

```diff
@@ -374,60 +392,83 @@ export class PlatformStore {
    *
    * In the end candidate with max weight is selected. However if candidate's
    * weight is equal to local platform, then local platform is selected.
-   * @return {!./subscription-platform.SubscriptionPlatform}
+   * @return {!Promise<!./subscription-platform.SubscriptionPlatform>}
+   * @param {string=} optionalFactor if present only use this factor for calculation
    * @private
    */
-  selectApplicablePlatform_() {
+  selectApplicablePlatform_(optionalFactor) {
     const localPlatform = this.getLocalPlatform();
     let localWeight = 0;
-    /** @type {!Array<!Object<!./subscription-platform.SubscriptionPlatform, number>>} */
-    const platformWeights = [];
 
     dev().assert(this.areAllPlatformsResolved_(),
         'All platforms are not resolved yet');
 
-    this.getAvailablePlatforms().forEach(platform => {
-      let weight = 0;
+    /// Subscriber wins immediatly.
+    const availablePlatforms = this.getAvailablePlatforms();
+    for (const platformIdx in availablePlatforms) {
+      const platform = availablePlatforms[platformIdx];
       const entitlement =
           this.getResolvedEntitlementFor(platform.getServiceId());
-
-      // Subscriber wins immediatly.
       if (entitlement.isSubscriber()) {
-        weight += 100000;
+        return Promise.resolve(platform);
       }
-
-      // Add the base score
-      weight += platform.getBaseScore();
-
-      // If supports the current viewer, gains weight 9
-      if (platform.supportsCurrentViewer()) {
-        weight += this.scoreConfig_['supportsViewer'];
-      }
-
-      platformWeights.push({
-        platform,
-        weight,
-      });
-      if (platform.getServiceId() === 'local') {
-        localWeight = weight;
-      }
-    });
-
-    platformWeights.sort(function(platform1, platform2) {
-      return platform2.weight - platform1.weight;
-    });
-    // Nobody supports current viewer, nor is anybody subscribed
-    if (platformWeights.length === 0) {
-      return localPlatform;
     }
 
-    const winningWeight = platformWeights[0].weight;
+    // Get weights for all of the platforms
+    return Promise.all(this.getAvailablePlatforms().map(platform => {
```

Comment:
I'd want this part of the code to be tested because it looks like something that can easily be broken if someone else goes in to change something. It might be easier to move this to a separate method for testing purposes.

Reply:
I broke it out and refactored how the weight sort is done - added a test for the new method.